Add a Q# quantum state visualizer

marshallsa opened the pull request:
This is a web-based debugger that can step (forward and back) through a quantum simulation, showing the operation call stack and quantum state at each step. It's based on a visualizer by Chris Granade.

Co-authored-by: Chris Granade <chgranad@microsoft.com>

anpaz-msft commented:
Looks pretty cool! Please do give me the change to review it. I'm on vacation but plan to spend some time on Thursday.
Thanks!

anpaz-msft commented on Samples/src/VisualDebugger/Hello.qs:
```diff
@@ -0,0 +1,60 @@
+namespace vis_sim {
```
Missing copyright notice.

anpaz-msft commented on Samples/src/VisualDebugger/Hubs.cs:
```diff
@@ -0,0 +1,23 @@
+using Microsoft.AspNetCore.SignalR;
+using System.Threading.Tasks;
+
+namespace vis_sim
```
We should probably put this sample under the `Microsoft.Quantum.Samples.VisualDebugger` namespace...

anpaz-msft commented on Samples/src/VisualDebugger/Hubs.cs:
```diff
@@ -0,0 +1,23 @@
+using Microsoft.AspNetCore.SignalR;
```
copyright notice.

anpaz-msft commented on Samples/src/VisualDebugger/Hubs.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace vis_sim
+{
+    internal class VisualHub : Hub
```
Let's add some ///summary with what this class does and how it works, for those not familiar with SignalR.

anpaz-msft commented on Samples/src/VisualDebugger/Program.cs:
```diff
@@ -0,0 +1,13 @@
+﻿using Microsoft.Quantum.Simulation.Simulators;
```
Copyright notice.

anpaz-msft commented on Samples/src/VisualDebugger/Program.cs:
```diff
@@ -0,0 +1,13 @@
+﻿using Microsoft.Quantum.Simulation.Simulators;
+
+namespace vis_sim
```
Namespace

anpaz-msft commented on Samples/src/VisualDebugger/Program.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace vis_sim
+{
+    internal class Program
```
Similarly, let's add some description about what this program does and how to call it from the command line.
A couple of troubleshooting tips would be nice (although those can go in the README.md)

marshallsa replied:
Did you run into any issues that you think should be mentioned in a troubleshooting section?

anpaz-msft commented on Samples/src/VisualDebugger/Startup.cs:
```diff
@@ -0,0 +1,26 @@
+using Microsoft.AspNetCore.Builder;
```
Copyright notice.

anpaz-msft commented on Samples/src/VisualDebugger/Startup.cs:
```diff
@@ -0,0 +1,26 @@
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace vis_sim
```
Namespace

anpaz-msft commented on Samples/src/VisualDebugger/Startup.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace vis_sim
+{
+    internal class Startup
```
At first glance, it's unclear about why there is a `Startup` and why it's not called from `Program.Main`. Adding some description here would be nice.

anpaz-msft commented on Samples/src/VisualDebugger/VisualDebugger.cs:
```diff
@@ -0,0 +1,118 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace vis_sim
```
Namespace.

anpaz-msft commented on Samples/src/VisualDebugger/VisualDebugger.cs:
```diff
@@ -0,0 +1,118 @@
+
+        public VisualDebugger(QuantumSimulator simulator)
+        {
+            simulator.OnOperationStart += OnOperationStartHandler;
```
You should check that simulator != null and throw `ArgumentNullException` at top.

anpaz-msft commented on Samples/src/VisualDebugger/VisualDebugger.cs:
```diff
@@ -0,0 +1,118 @@
+
+        public async Task ReplayHistory(IClientProxy client)
+        {
+            foreach (var (method, args) in history)
```
Can you please add brackets to wrap the foreach?

anpaz-msft commented on Samples/src/VisualDebugger/VisualDebugger.cs:
```diff
@@ -0,0 +1,118 @@
+                await client.SendCoreAsync(method, args);
+        }
+
+        private T GetService<T>() => (T) host.Services.GetService(typeof(T));
```
By internal convention, the body after `=>` should be on the next line.

anpaz-msft commented on Samples/src/VisualDebugger/VisualDebugger.cs:
```diff
@@ -0,0 +1,118 @@
+            advanceEvent.WaitOne();
+        });
+
+        private void OnOperationStartHandler(ICallable operation, IApplyData arguments)
```
Chris added recently the `OnQubitsAllocation` (or similar) to the `SimulatorBase`. You should change the code here to leverage it.

anpaz-msft commented on Samples/src/VisualDebugger/package.json:
```diff
@@ -0,0 +1,31 @@
+{
```
I'm under the impression that `package-lock.json` should not be committed, but I could be wrong.

marshallsa replied:
It looks like the npm documentation says it's intended to be committed as a snapshot of all the package versions in node_modules: https://docs.npmjs.com/files/package-lock.json

anpaz-msft commented on Samples/src/VisualDebugger/webpack.config.js:
```diff
@@ -0,0 +1,37 @@
+const path = require("path");
```
I'm not convinced that using `webpack` give us any benefit. We literally have 3 files (one .ts, one .html and one .css), adding an extra build step to pack this seems too much work for what we need.

marshallsa replied:
I think we need something like Webpack or Browserify to make the npm modules we import work in the browser (`@aspnet/signalr` and `chart.js`). I tried compiling directly with `tsc` and it didn't work.

anpaz-msft commented on Samples/src/VisualDebugger/websrc/app.ts:
```diff
@@ -0,0 +1,304 @@
+            {
+                data: [],
+                label: "Real",
+				backgroundColor: "#ff0000",
```
The file has some tabs , please convert all to spaces.

anpaz-msft commented on Samples/src/VisualDebugger/websrc/app.ts:
```diff
@@ -0,0 +1,304 @@
+    }
+});
+
+function updateChart(state: State) {
```
It would be nice if the x-axis showed the base vector as the label, is that possible?

marshallsa replied:
Yes, they should be there - are they not working? This is what I see:

![Basis Vectors](https://user-images.githubusercontent.com/33814365/63982986-9cadcc00-ca79-11e9-8dd2-7e1d084b0d81.png)

anpaz-msft replied:
Ah! You're right. I was expecting them to be in the x-axis line (i.e. on the `0`); further more, they actually do show but they are mostly covered by the scrollbar:

![image](https://user-images.githubusercontent.com/24311286/64479603-54c03080-d16e-11e9-923c-ca7f52d9b677.png)

It's ok if we leave them at the bottom, but could we push them a little higher so they are not blocked by the scrollbar?

Further, allocating new qubits is not showing the new basis state after the allocation, i.e.:
![image](https://user-images.githubusercontent.com/24311286/64479660-d7e18680-d16e-11e9-8450-dfc42abb2ffb.png)

Is that something we could add?

marshallsa replied:
Hmm, I don't see the bottom scrollbar even when I make my window small. What browser are you using?

marshallsa replied:
It seems like the `OnAllocateQubits` handler is called just before the qubits are actually allocated - so the state that I can get in the handler is out of date. For example after the third allocation, the state would show only 4 basis vectors, but after the first H it would show 8 vectors.

@cgranade Is it possible for the simulator to invoke `OnAllocateQubits` after calling `manager.Allocate`?

anpaz-msft replied:
@marshallsa, OnOperationStart is called *before* the operation is called, `onAllocate` should remain consistent with that.
Are you saying that before calling 'H', there are only 4 basis state? that seems odd, there might be something else going on, by the time 'H' is highlighted (i.e. after `Allocate` is called) there should already be 8 basis state vectors.

marshallsa replied:
No, I think you're right that there are 8 basis vectors by the time `OnOperationStart` is called for H, but I was trying to show all 8 of them on the last "allocate 1 qubit" line instead of the first "H" line. Do you prefer showing 4 basis states on the last allocate line and then at the start of the first H line (before H actually runs), it shows 8?

anpaz-msft replied:
I think it's ok to be consistent, so when the last Allocation is highlighted then there should be 4, when H gets highlighted there should be 8. Again, the instruction listed is what is going to be executed, not what was already executed.

marshallsa replied:
That makes sense - allocation should work that way now. I also changed the green checkmark to a blue arrow like the rest of the operations so it's clear that the allocation is the next thing to execute. Sorry for the confusion!

marshallsa commented:
@anpaz-msft Thanks for your review! I think I've addressed all of your comments (except for a few questions I had). Could you take another look when you have time?

anpaz-msft commented:
One more comment: calling this a Visual Debugger is a bit of a stretch as you can't do most of the tasks normally associated with a debugger (check variables, put breakpoints, evaluate expressions, etc). Maybe we should call it "QuantumStateVisualizer" or just "StateVisualizer" instead?

anpaz-msft commented:
You're going to hate me, but for some reason the dotnet process keeps ligering around once you stop using Ctrl+C. My experience is that this is typically that you're starting a thread and continue execution on Main, Main finishes but execution continues due to the open thread, but its' just an idea.

marshallsa commented:
@anpaz-msft Ctrl+C should work now. :)

Do you still have the bug where the scrollbar at the bottom covers up the basis vector labels? I haven't been able to reproduce that.

anpaz-msft commented on Samples/src/StateVisualizer/websrc/css/main.css:
```diff
@@ -0,0 +1,83 @@
+    margin: 0;
+    padding: 0;
+    height: 100%;
+    width: 100%;
```
```suggestion
    width: 100%;
    overflow: hidden;
```

anpaz-msft replied:
this fixes the superfluos scrollbar that is hidden the base states.

anpaz-msft commented on Samples/src/StateVisualizer/package-lock.json:
```diff
@@ -0,0 +1,5366 @@
+{
+  "name": "vis-sim",
```
```suggestion
  "name": "qs-state-visualizer",
```

anpaz-msft replied:
that's what it generated on my laptop... again, I don't think this should be committed :)

marshallsa replied:
Oops, I forgot to update package-lock.json after renaming the package. If you want to add it to gitignore I am also OK with that. :)

marshallsa commented:
> One very minor thing.
> Should I merge it after it?

Sure, I think it should be ready to merge (unless you want to remove package-lock.json first).

anpaz-msft commented:
Thanks for incorporating all the feedback. Great contribution!
